src/ray/core_worker/core_worker.cc

```diff
@@ -75,62 +75,216 @@ void GroupObjectIdsByStoreProvider(const std::vector<ObjectID> &object_ids,
+              ? ComputeDriverIdFromJob(options_.job_id)
+              : (options_.num_workers == 1 ? WorkerID::FromRandom() : WorkerID::Nil())) {
+  // Initialize logging if log_dir is passed. Otherwise, it must be initialized
+  // and cleaned up by the caller.
+  if (options_.log_dir != "") {
+    std::stringstream app_name;
+    app_name << LanguageString(options_.language) << "-core-"
+             << WorkerTypeString(options_.worker_type);
+    if (!global_worker_id_.IsNil()) {
+      app_name << "-" << global_worker_id_;
+    }
+    RayLog::StartRayLog(app_name.str(), RayLogLevel::INFO, options_.log_dir);
+    if (options_.install_failure_signal_handler) {
+      RayLog::InstallFailureSignalHandler();
+    }
+  }
+
+  RAY_CHECK(options_.num_workers > 0);
+  if (options_.worker_type == WorkerType::DRIVER) {
+    // Driver process can only contain one worker.
+    RAY_CHECK(options_.num_workers == 1);
+  }
+
+  RAY_LOG(INFO) << "Constructing CoreWorkerProcess, pid: " << getpid();
+
+  if (options_.num_workers == 1) {
+    // We need to create the worker instance here if:
+    // 1. This is a driver process. In this case, the driver is ready to use right after
+    // the CoreWorkerProcess::Initialize.
+    // 2. This is a Python worker process. In this case, Python will invoke some core
+    // worker APIs before `CoreWorkerProcess::StartExecutingTasks` is called. So we need
+    // to create the worker instance here. One example of invocations is
+    // https://github.com/ray-project/ray/blob/45ce40e5d44801193220d2c546be8de0feeef988/python/ray/worker.py#L1281.
+    if (options_.worker_type == WorkerType::DRIVER ||
+        options_.language == Language::PYTHON) {
+      CreateWorker();
+    }
+  }
+}
+
+CoreWorkerProcess::~CoreWorkerProcess() {
+  RAY_LOG(INFO) << "Core worker process is exiting.";
+  {
+    absl::MutexLock lock(&worker_map_mutex_);
+    RAY_CHECK(workers_.empty());
+  }
+  if (options_.log_dir != "") {
+    RayLog::ShutDownRayLog();
+  }
+}
+
+void CoreWorkerProcess::EnsureInitialized() {
+  RAY_CHECK(instance_) << "The core worker process is not initialized yet or already "
+                       << "shutdown.";
+}
+
+CoreWorker &CoreWorkerProcess::GetCoreWorker() {
+  EnsureInitialized();
+  if (instance_->options_.num_workers == 1) {
+    return *instance_->global_worker_;
+  }
+  auto ptr = current_core_worker_.lock();
+  RAY_CHECK(ptr != nullptr)
+      << "The current thread is not bound with a core worker instance.";
+  return *ptr;
+}
+
+void CoreWorkerProcess::SetCurrentThreadWorkerId(const WorkerID &worker_id) {
+  EnsureInitialized();
+  if (instance_->options_.num_workers == 1) {
+    RAY_CHECK(instance_->global_worker_->GetWorkerID() == worker_id);
+    return;
+  }
+  current_core_worker_ = instance_->GetWorker(worker_id);
+}
+
+std::shared_ptr<CoreWorker> CoreWorkerProcess::GetWorker(
+    const WorkerID &worker_id) const {
+  absl::MutexLock lock(&worker_map_mutex_);
```

Comment:
probably better to have a read-write lock here, since after the workers are initialized, almost all the access to this map are read, and they don't need to be exclusive.

Reply:
Nice catch! Done.